torch/utils/_pytree.py

```diff
@@ -190,9 +190,9 @@ def tree_unflatten(values: List[Any], spec: TreeSpec) -> PyTree:
 
     return unflatten_fn(child_pytrees, spec.context)
 
-def tree_map(fn: Any, pytree: PyTree) -> PyTree:
+def tree_map(fn: Any, pytree: PyTree, *rest: Any) -> PyTree:
     flat_args, spec = tree_flatten(pytree)
-    return tree_unflatten([fn(i) for i in flat_args], spec)
+    return tree_unflatten([fn(*flat_arg) for flat_arg in zip(flat_args, *rest)], spec)
```

Comment:
This doesn't look correct to me. Every pytree in `rest` should have the same tree-structure as `pytree`. I would have expected the code to look like:
```
flat_args, spec = tree_flatten(pytree)
flat_rests = []
for r in rest:
   flat_r, r_spec = tree_flatten(r)
   check_same(r_spec, spec)
   flat_rests.append(r)
   
tree_unflatten([fn(*flat_arg) for flat_arg in zip(flat_args, *flat_rests)], spec)
```
```

Reply:
We may need to implement a _"flatten up to"_ method for treespecs. It will keep the arbitrary subtree as it is rather than flatten the whole tree.

`flatten_up_to` will produce a list of subtrees (a single leaf is also a tree) with the same size of the referenced tree.

```python
_, spec = tree_flatten({'a': 0, 'b': 1})
spec.flatten_up_to({'a': {'c': 0, 'd': 1}, 'b': [2, 3, 4]})
# [{'c': 0, 'd': 1}, [2, 3, 4]]
tree_flatten({'a': {'c': 0, 'd': 1}, 'b': [2, 3, 4]})[0]
# [0, 1, 2, 3, 4]
```

```python
def tree_map(func, tree, *rests):
    leaves, treespec = tree_flatten(tree)
    flat_args = [leaves] + [treespec.flatten_up_to(r) for r in rests]
    flat_results = map(func, *flat_args)
    return tree_unflatten(flat_results, treespec)
```

**NOTE: be careful of the `dict` key order.** You cannot simply flatten the two trees (or more), and then map and unflatten back. JAX internally sorts the `dict` key order in `PyTreeDef` which is not documented yet ([JAX: Pytrees](https://jax.readthedocs.io/en/latest/pytrees.html)). See also [OpTree: Key Ordering for Dictionaries](https://github.com/metaopt/optree#key-ordering-for-dictionaries).

We should implement the `flatten_up_to` method to handle the key order to produce the same order of values as the referenced treespec. Otherwise, the multi-treemap would have undefined behavior.

```python
In [1]: import torch.utils._pytree as pytree

In [2]: tree1 = {'a': 0, 'b': 1}

In [3]: tree2 = {'b': 1, 'a': 0}

In [4]: leaves1, spec1 = pytree.tree_flatten(tree1)

In [5]: leaves2, spec2 = pytree.tree_flatten(tree2)

In [6]: leaves1
Out[6]: [0, 1]

In [7]: leaves2
Out[7]: [1, 0]

In [8]: spec1
Out[8]: TreeSpec(dict, ['a', 'b'], [*, *])

In [9]: spec2
Out[9]: TreeSpec(dict, ['b', 'a'], [*, *])

In [10]: tree1 == tree2
Out[10]: True

In [11]: leaves1 == leaves2
Out[11]: False

In [12]: spec1 == spec2
Out[12]: False
```

Reply:
We already have an implementation for `broadcast_to_and_flatten`:

https://github.com/pytorch/pytorch/blob/34e8eb229db76f7f5eb8f18c062dbd1ee47f8b12/torch/utils/_pytree.py#L293-L328

The _"flatten up to"_ would be similar to this. Flatten only one level and then recursion. Both `flatten_up_to` and `broadcast_to_and_flatten` need to add some checks for `dict.keys()` equality.

Some references may be useful:

[`optree.tree_map`](https://github.com/metaopt/optree/blob/8bb88014e370d007a60e86a6828dfdfd092e0a28/optree/ops.py#L414-L467)
[`optree.broadcast_prefix`](https://github.com/metaopt/optree/blob/8bb88014e370d007a60e86a6828dfdfd092e0a28/optree/ops.py#L1000-L1061) (which is built on top of `tree_map`)